lib/SILOptimizer/LoopTransforms/LoopBranchHoist.cpp

```diff
@@ -0,0 +1,321 @@
+  SILBuilder(clonedCondBr).createBranch(clonedCondBr->getLoc(), clonedBB[falseBB]);
+  clonedCondBr->eraseFromParent();
+
+  // And then remove the boolean phi from the header in each half.
+  header->eraseArgument(argNumber);
+  clonedHeader->eraseArgument(argNumber);
+
+  // But preheader becomes conditional to select which loop initially.
+  redirectToBothLoops(preheader, argNumber, header, clonedHeader);
+
+  for (auto latch : latchBlocks) {
+    auto terminator = latch->getTerminator();
+    auto arg = terminator->getAllOperands()[argNumber].get()->getDefiningInstruction();
+    bool trueDom = domTree->dominates(trueBBFront, arg);
+    bool falseDom = domTree->dominates(falseBBFront, arg);
+
+    auto clonedLatch = clonedBB[latch];
+    if (latch == clonedLatch || !clonedLatch) {
+      // No clone so latch is just in the true loop
+      if (trueDom) {
+        redirectToBothLoops(latch, argNumber, header, clonedHeader);
+      } else {
+        redirectToSingleLoop(latch, argNumber, header);
+      }
+    } else {
+      if (falseDom) // false loop changed the loop value
+        redirectToBothLoops(clonedLatch, argNumber, header, clonedHeader);
+      else
+        redirectToSingleLoop(clonedLatch, argNumber, clonedHeader);
+
+      bool onlyInFalse = domTree->dominates(falseBBFront, &latch->front());
+      if (onlyInFalse) {
+        latch->eraseFromParent();
+      } else {
+        if (trueDom) // true loop changed the loop value
+          redirectToBothLoops(latch, argNumber, header, clonedHeader);
+        else
+          redirectToSingleLoop(latch, argNumber, header);
+      }
+    }
+  }
+  return true;
+}
+
+// =============================================================================
+//                                 Driver
+// =============================================================================
+
+namespace {
+
+  class LoopBranchHoisting : public SILFunctionTransform {
+
+    void run() override {
+      bool Changed = false;
+
+      auto *Fun = getFunction();
+      auto I1Ty = SILType::getBuiltinIntegerType(1, Fun->getASTContext());
+
+      SILLoopInfo *LoopInfo = PM->getAnalysis<SILLoopAnalysis>()->get(Fun);
+      DominanceAnalysis *DA = PM->getAnalysis<DominanceAnalysis>();
+      DominanceInfo *domTree = nullptr;
+
+      // Collect innermost loops.
+      SmallVector<SILLoop *, 16> InnermostLoops;
+      for (auto *Loop : *LoopInfo) {
+        SmallVector<SILLoop *, 8> Worklist;
+        Worklist.push_back(Loop);
+
+        for (unsigned i = 0; i < Worklist.size(); ++i) {
+          auto *L = Worklist[i];
+          for (auto *SubLoop : *L)
+            Worklist.push_back(SubLoop);
+          if (L->getSubLoops().empty())
+            InnermostLoops.push_back(L);
+        }
+      }
+
+      // Check innermost loops.
+      for (auto *loop : InnermostLoops) {
+        auto *header = loop->getHeader();
+        unsigned int headerArgs = header->getNumArguments();
```

Comment:
LLVM style is to just use unsigned. The int is redundant. Another thing you can use here to make this cleaner is to use the range helper. Then you get:

```
for (unsigned argNumber : range(headerArgs)) {
...
}
```

Can you also invert the if statements to use early exits? I.e.:

```
if (!header->getArgument(i)->getType() == I1Ty)
  continue;
if (!domTree) {
  domTree = DA->get(Fun);
}
if (!tryToHoist(...))
  continue;
Changed = true;
break;
```